File: Algorithm_148_Sort List.java
// 思路：使用快慢指针找到链表的中间位置，并将前一半的最后一个节点的next置为null，从而从中间切断链表
// 并对左右两半部分递归调用sortList，最后使用归并方法合并成一个链表

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode prev = null, slow = head, fast = head;
        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        prev.next = null; // cut the list
        ListNode n1 = sortList(head);
        ListNode n2 = sortList(slow);
        return merge(n1, n2);
    }

    public ListNode merge(ListNode n1, ListNode n2) {
        if (n1 == null) {
            return n2;
        } else if (n2 == null) {
            return n1;
        }
        ListNode start = new ListNode(0), cur = start;
        while (n1 != null && n2 != null) {
            if (n1.val < n2.val) {
                cur.next = n1;
                n1 = n1.next;
            } else {
                cur.next = n2;
                n2 = n2.next;
            }
            cur = cur.next;
        }
        if (n1 == null) {
            cur.next = n2;
        } else {
            cur.next = n1;
        }
        return start.next;
    }
}

File: Algorithm_148_Sort List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        slow = fast = ListNode(0)
        slow.next = head
        # get to the middle node of the list
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        head2 = slow.next
        slow.next = None
        newHead1 = self.sortList(head)
        newHead2 = self.sortList(head2)
        return self.merge(newHead1, newHead2)
    
    def merge(self, head1: ListNode, head2: ListNode) -> ListNode:
        if not head1:
            return head2
        elif not head2:
            return head1
        cur = dummy = ListNode(0)
        cur1, cur2 = head1, head2
        while cur1 and cur2:
            if cur1.val < cur2.val:
                cur.next = cur1
                cur1 = cur1.next
            else:
                cur.next = cur2
                cur2 = cur2.next
            cur = cur.next
        cur.next = cur2 if not cur1 else cur1
        return dummy.next